backend/src/v5/routes/sso/aad/aad.js

```diff
@@ -0,0 +1,165 @@
+/**
+ *  Copyright (C) 2022 3D Repo Ltd
+ *
+ *  This program is free software: you can redistribute it and/or modify
+ *  it under the terms of the GNU Affero General Public License as
+ *  published by the Free Software Foundation, either version 3 of the
+ *  License, or (at your option) any later version.
+ *
+ *  This program is distributed in the hope that it will be useful,
+ *  but WITHOUT ANY WARRANTY; without even the implied warranty of
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+const { checkAadConfig, getUserDetailsAndValidateEmail } = require('../../../middleware/dataConverter/inputs/sso/aad');
+const { Router } = require('express');
+const Users = require('../../../processors/users');
+const { getAuthenticationCodeUrl } = require('../../../utils/sso/aad');
+const { respond } = require('../../../utils/responder');
+const { templates } = require('../../../utils/responseCodes');
+const { validateSignUpData } = require('../../../middleware/dataConverter/inputs/users');
+
+const authenticateRedirectUri = 'http://localhost/api/v5/sso/aad/authenticate-post';
+const signupRedirectUri = 'http://localhost/api/v5/sso/aad/signup-post';
+// Frontend uri for signup form
+const sigupUri = '';
+
+const authenticate = (req, res) => {
+	try {
+		const params = { redirectUri: authenticateRedirectUri, state: sigupUri };
+		const authenticationCodeUrl = getAuthenticationCodeUrl(params);
```

Comment:
I'm pretty sure this is a function that returns a promise, so you're redirecting the response to a promise...?

![image](https://user-images.githubusercontent.com/11945337/179238499-fd8f774e-d818-4da6-ac3b-8ff5829718d9.png)

Reply:
also I would put `getAuthenticationCodeUrl` in services  once again and expose it as a function